### ISSUE TITLE
Exception at interleaved -> interleaved processing

### ISSUE THREAD

**Issue by Gobra** [halide/Halide#3080]
Here's quite simple code that fails on the `realize` call with error message `Constraint violated: f0.stride.0 (3) == 1 (1)`
```
Halide::Var x, y, c;
Halide::Func processor;
auto input = Halide::Buffer<uint8_t>::make_interleaved((uint8_t *)bitmap.bits(), bitmap.width(), bitmap.height(), 3);
auto output = Halide::Buffer<uint8_t>::make_interleaved(bitmap.width(), bitmap.height(), 3);

processor(x, y, c) = input(x, y, c);
processor.realize(output);
```

To me the code makes perfect sense as both input and output are of the same dimensions and layout.
I've tried "rendering" to the input filter with `processor.realize(input);` and it fails with the same error message.

Changing output to default (planar?) layout with auto `output = Halide::Buffer<uint8_t>(bitmap.width(), bitmap.height(), 3);` makes it working.

Is that a bug or did I miss something obvious in documentation regarding the data representation (layout)?

Should it be important - I'n running with Win10 x64, llvm 6.0, MSVC2017, Halide "master" branch from 21 june 2018.

**Comment by abadams** [comment]
Take a look at http://halide-lang.org/tutorials/tutorial_lesson_16_rgb_generate.html

By default halide assumes the first dimension (x in this case) is dense in memory (stride 1). This is so it can generate good dense vector loads. You can tell it otherwise using processor.output_buffer().dim(0).set_stride(3)

**Comment by Gobra** [comment]
Yes, but I have created input buffer as `make_interleaved`, it shows stride 3 for x and 3072 for y (buffer width is 1024) which seems to be right. As output buffer is set the same way, why would Halide assume difference stride for x from one defined for both input and target output?

**Comment by abadams** [comment]
"processor" is compiled without knowledge of the output buffer you're going to use. You could call realize with any argument, and it would only compile once. I don't remember if this is also true of input buffers. I think when you use them directly perhaps Halide inspects the layout. If it were an ImageParam you'd have the same problem on the input as you do on the output.

**Comment by Gobra** [comment]
Thanks for your help, it makes sense and with your suggested fix it works now.

It's still quite confusing for a newbie and I'm sure a lot of programmer might fall to the same "trap". I've read tutorial 16 before posting the issue, but it also a bit confusing due to those differences between JIT and AOT schemes.

May I suggest maybe adding a page to the project wiki about buffer memory layouts and ways to treat it with Halide API? The most basic examples for JIT/AOT in addition to the tutorial 15 and 16 should be of big help to those only jumping onto the ship.

**Comment by SuTanTank** [comment]
I encountered the same problem when learning using Halide. 
I was trying to process a Buffer created from cv::Mat, which is interleaved.

**Comment by SuTanTank** [comment]
I found that for a 3-channel interleaved image, `reorder(c, x, y)` cause significant slower process than doing nothing, which seem not making sense. Why following actual layout (y>x>c) is worse than (c>y>x) ?

**Comment by liupeiqiHN** [comment]
I gave up using Halide's AOT, there were too many puzzling problems, and algorithms that worked fine on JIT would cause the program to crash under AOT. I don't know how to handle the conversion relationship between cv::Mat and Halide::Buffer,  there are always a lot of hidden problems here, it would be nice to have a robust conversion method.

**Comment by abadams** [comment]
> [!NOTE]
> **Automated triage:** this issue was flagged by Claude as likely safe to close, for the
> reason below. The reasoning was checked against the current source, and a second pass
> tried and failed to refute it, but no human verified this specific text.
> **If the reason is wrong, please reopen the issue and say so** — some fraction of these
> are expected to be wrong.

**Reason (CLOSE_WONTFIX):** Not a bug: a JIT-realized Func is compiled independently of the buffer passed to realize(), so its output defaults to stride-1 in x and an interleaved output violates that constraint. abadams explained this and the reporter confirmed output_buffer().dim(0).set_stride(3) fixes it. The residual ask is an open-ended 'write a wiki page about layouts', which tutorial lesson 16 and the Buffer docs already cover.

**Evidence:** Documented at src/runtime/HalideBuffer.h:1925-1930 ('Passing it to a generator requires that the generator has been compiled with support for interleaved ... memory layouts'); tutorial/lesson_16_rgb_generate.cpp covers exactly this. Thread resolved 2018-06-23.

Closing - this is expected behavior, not a bug: a Func is compiled without knowledge of the buffer passed to realize(), and defaults to stride 1 in the innermost dimension. Declare the layout with output_buffer().dim(0).set_stride(3) (see tutorial lesson 16 and the make_interleaved docs in HalideBuffer.h). The follow-up documentation request is too open-ended to track here.
